### PR TITLE
Introduced colored backpack calendar

### PR DESCRIPTION
**Pull request by Dominguezzz**
Backpack
- removed error-prone timezone dependent Calendar from BpkCalendarController api(startDate and endDate). Now CalendarDay is used similarly to selectedDay and CalendarRange.
- introduced CalendarColoring object that holds the coloring information
- introduced CalendarDrawingParams instead of passing map and relying on keys
- cleaned up some method parameters: use CalendaDay instead of passing year: Int, month: Int, day: Int
- refactored CalendarDay to be a data class
	- removed unnecessary constructors
	- made it immutable
	- removed Calendar api from its interface (now it's an in class dependency that can be refactored later)
	- removed SimpleDateFormat from toString for performance
- MonthView
	- removed unnecessary tag handling
	- fixed an edge case bug when the range of the last day of month did not render properly on selection
	- drawing of the small colored dots based on CalendarDrawingParams.calendarColoring
	- same date selection text color and filled circle changes
	- selected date color and size changes
	- range lane height and color changes
	- today is not a circle but colored label
	- removed dead code 

Example app:
- new switch to control colored/non-colored Calendar
- dummy data generation for colored calendar

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/master/CONTRIBUTING.md)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt [hunk 1]
```diff
@@ -70,7 +73,14 @@ abstract class BpkCalendarController {
...
+    val DEFAULT_END_DATE: CalendarDay = Calendar.getInstance()
+      .apply {
+        add(Calendar.YEAR, 1)
+        add(Calendar.DATE, -1)
+      }
+      .toCalendarDay()
   }
 }
+
+fun Calendar.toCalendarDay() = CalendarDay(year = get(Calendar.YEAR), month = get(Calendar.MONTH), day = get(Calendar.DAY_OF_MONTH))
```
Comment (amolgupta) [inline Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt]: this can be an additional constructor in `CalendarDay` class as 

``` 
constructor(day: Calendar) : this(day.get(Calendar.YEAR), day.get(Calendar.MONTH), day.get(Calendar.DAY_OF_MONTH))
```
as it is the responsibility of that class and is more discoverable to the consumer.

File: Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt [hunk 2]
```diff
@@ -29,15 +30,6 @@ class BpkCalendarControllerTest {
     subject = BpkCalendarControllerTestImpl()
   }
 
-  @Test
```
Comment (amolgupta) [inline Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt]: Why is this removed if the default date is still there?

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/MonthAdapter.kt [hunk 3]
```diff
@@ -68,24 +64,14 @@ internal class MonthAdapter(
...
-
-        drawingParams.clear()
-
-        val positionWithStart = position + controller.startDate.get(Calendar.MONTH)
+        val positionWithStart = position + controller.startDate.month
         val month = positionWithStart % MONTHS_IN_YEAR
-        val year = positionWithStart / MONTHS_IN_YEAR + controller.startDate.get(Calendar.YEAR)
-        val day = if (isSelectedDayInMonth(selectedDay, year, month)) selectedDay.day else -1
+        val year = positionWithStart / MONTHS_IN_YEAR + controller.startDate.year
+        val selectedDay = if (selectedDay != null && (isSelectedDayInMonth(selectedDay!!, year, month))) selectedDay!!.day else null
```
Comment (amolgupta) [inline Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/MonthAdapter.kt]:
```suggestion
val selectedDay = selectedDay?.let {
      if (isSelectedDayInMonth(it, year, month)) it.day else null
    }
```

File: app/src/main/res/layout/fragment_calendar.xml [hunk 4]
```diff
@@ -1,12 +1,21 @@
 <?xml version="1.0" encoding="utf-8"?>
-<FrameLayout
+<LinearLayout
```
Comment (amolgupta) [inline app/src/main/res/layout/fragment_calendar.xml]: We generally make a new story for a different variant of a component to make it easy for the developers to refer to the usage. Please see `ComponentRegistry.kt` for an example

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt [hunk 5]
```diff
@@ -1,76 +1,58 @@
...
+  companion object {
+    fun of(timeInMillis: Long): CalendarDay {
+      val utcCalendar = utcCalendar()
+      utcCalendar.timeInMillis = timeInMillis
+      return utcCalendar.toCalendarDay()
+    }
+
+    fun today(): CalendarDay = of(System.currentTimeMillis())
+
+    fun utcCalendar() = Calendar.getInstance(TimeZone.getTimeZone("UTC"))!!
```
[the remark this answers is not in the document]
Comment (Dominguezzz) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt]: 👍

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt [hunk 6]
```diff
@@ -1,76 +1,58 @@
...
+  private fun Calendar.trimCalendar() = apply {
+    set(Calendar.HOUR_OF_DAY, 0)
+    set(Calendar.MINUTE, 0)
+    set(Calendar.SECOND, 0)
+    set(Calendar.MILLISECOND, 0)
+  }
+
+  override fun toString(): String = String.format("%04d-%02d-%02d", year, month + 1, day)
+
+  fun newInstanceByAddedDays(days: Int): CalendarDay {
```
[the remark this answers is not in the document]
Comment (Dominguezzz) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt]: addDays might indicate a mutable apo

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt (hunk 1 above)
Comment (Dominguezzz) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt]: Calendar is a bad, mutable api that I intentionally removed from the CalendarDay api to keep it clean. I moved it to call site so it's easier to be removed.

File: Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt (hunk 2 above)
Comment (Dominguezzz) [reply Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt]: There is no default date anymore just a data class 
```
CalendarDay(
  val year: Int,
  val month: Int,
  val day: Int
)
```

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/MonthAdapter.kt (hunk 3 above)
Comment (Dominguezzz) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/MonthAdapter.kt]: 👍

File: app/src/main/res/layout/fragment_calendar.xml (hunk 4 above)
Comment (Dominguezzz) [reply app/src/main/res/layout/fragment_calendar.xml]: Cool, I will move it to a new story.

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (tiagoengel) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt]: I think the fact it returns a CalendarDay is enough to tell it returns a new instance and is not mutating.

File: Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt (hunk 2 above)
Comment (tiagoengel) [reply Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt]: But isn't this testing the default start and end date in the controller? Afaik those are still there. (DEFAULT_START_DATE and DEFAULT_END_DATE)

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt (hunk 1 above)
Comment (tiagoengel) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt]: Makes sense. Can we make `toCalendarDay` internal to make sure is not used outside our lib?

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dominguezzz) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/model/CalendarDay.kt]: 👍

File: Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt (hunk 1 above)
Comment (Dominguezzz) [reply Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt]: 👍

File: Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt (hunk 2 above)
Comment (Dominguezzz) [reply Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt]: Yes you are right. We are testing if we set a constant to a value then the value = constant. What value does this test bring?

Comment (tiagoengel) [reply Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt]: Well it's testing if the default start and end dates are what we expect. For one that test ensures that if for any reason the code is changed in a way that affects the default dates it will fail and prevent that error.

Plus it tests the default dates are start = today and end = today + 1 year

File: app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt [hunk 7]
```diff
@@ -88,7 +89,8 @@ object ComponentRegistry {
         "With divider without padding" story NodeData { Story of R.layout.fragment_card_with_divider_no_padding },
         "With divider and corner style large" story NodeData { Story of R.layout.fragment_card_with_divider_cornerstyle_large }
       )),
-    "Calendar" story NodeData { CalendarStory of R.layout.fragment_calendar },
+    "Calendar-Non-colored" story NodeData { NonColoredCalendarStory of R.layout.fragment_calendar_non_colored },
```
Comment (amolgupta) [inline app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt]:
```suggestion
   "Calendar" story NodeData({ children -> SubStory of children },
      mapOf(
        "Default" story NodeData { NonColoredCalendarStory of R.layout.fragment_calendar_non_colored },
        "Colored" story NodeData { ColoredCalendarStory of R.layout.fragment_calendar_colored }
      )),
```

File: app/src/main/res/layout/fragment_calendar_non_colored.xml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (amolgupta) [inline app/src/main/res/layout/fragment_calendar_non_colored.xml]: These two files are exactly the same and can just be `fragment_calendar.xml`

File: app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt (hunk 7 above)
Comment (Dominguezzz) [reply app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt]: 👍

File: Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt (hunk 2 above)
Comment (Dominguezzz) [reply Backpack/src/androidTest/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarControllerTest.kt]: 👍 If you find it useful, I put it back.

File: app/src/main/res/layout/fragment_calendar_non_colored.xml (hunk 8 above)
Comment (Dominguezzz) [reply app/src/main/res/layout/fragment_calendar_non_colored.xml]: 👍
